synapse/handlers/oidc_handler.py

```diff
@@ -0,0 +1,659 @@
+        finish_request(request)
+
+    async def handle_oidc_callback(self, request) -> None:
+        """Handle an incoming request to /_synapse/oidc/callback
+
+        Since we might want to display OIDC-related errors in a user-friendly
+        way, we don't raise SynapseError from here. Instead, we call
+        ``self._render_error`` which displays an HTML page for the error.
+
+        Most of the OpenID Connect logic happens here:
+
+          - first, we check if there was any error returned by the provider and
+            display it
+          - then we fetch the session cookie, decode and verify it
+          - the ``state`` query parameter should match with the one stored in the
+            session cookie
+          - once we known this session is legit, exchange the code with the
+            provider using the ``token_endpoint`` (see ``_exchange_code``)
+          - once we have the token, use it to either extract the UserInfo from
+            the ``id_token`` (``_parse_id_token``), or use the ``access_token``
+            to fetch UserInfo from the ``userinfo_endpoint``
+            (``_fetch_userinfo``)
+          - map those UserInfo to a Matrix user (``_map_userinfo_to_user``) and
+            finish the login
+
+        Args:
+            request (SynapseRequest): the incoming request from the browser.
+        """
+
+        # The provider might redirect with an error.
+        # In that case, just display it as-is.
+        if b"error" in request.args:
+            error = request.args[b"error"][0].decode()
+            description = request.args.get(b"error_description", [b""])[0].decode()
+            self._render_error(request, error, description)
+            return
+
+        # Fetch the session cookie
+        session = request.getCookie(SESSION_COOKIE_NAME)
+        if session is None:
+            self._render_error(request, "missing_session", "No session cookie found")
+            return
+
+        # Remove the cookie. There is a good chance that if the callback failed
+        # once, it will fail next time and the code will already be exchanged.
+        # Removing it early avoids spamming the provider with token requests.
+        request.addCookie(
+            SESSION_COOKIE_NAME,
+            b"",
+            path="/_synapse/oidc",
+            expires="Thu, Jan 01 1970 00:00:00 UTC",
+            httpOnly=True,
+        )
+
+        # Check for the state query parameter
+        if b"state" not in request.args:
+            self._render_error(request, "invalid_request", "State parameter is missing")
+            return
+
+        state = request.args[b"state"][0].decode()
+
+        # Deserialize the session token and verify it.
+        # FIXME: macaroon verification should be refactored somewhere else
+        try:
+            macaroon = pymacaroons.Macaroon.deserialize(session)
+        except MacaroonDeserializationException as e:
+            self._render_error(request, "invalid_session", str(e))
```

Comment:
Is this going to cause an end-user to see a Python exception? There's a few other places that an exception is directly converted to a string, so this goes for those as well.

Reply:
Haven't tried that, but probably. It's sometimes hard to write meaningful human-readable error messages, especially as I'm not a native english speaker. That's why I sometime got lazy for some of those messages that should probably never happen 😛 

I'll try to come up with something for those.

Reply:
It might be OK to have a fallback message that gets used in some places ("Something went wrong, retry." sort of thing.)